packages/records/lib/models/records.ts

```diff
@@ -216,53 +216,97 @@ export async function upsert({
-                if (delta !== 0) {
-                    await trx
-                        .from(RECORD_COUNTS_TABLE)
-                        .insert({
-                            connection_id: connectionId,
-                            model,
-                            environment_id: environmentId,
-                            count: delta
-                        })
-                        .onConflict(['connection_id', 'environment_id', 'model'])
-                        .merge({
-                            count: trx.raw(`${RECORD_COUNTS_TABLE}.count + EXCLUDED.count`),
-                            updated_at: trx.fn.now()
-                        });
+                // we need to know which records were updated, deleted, undeleted or unchanged
+                // we achieve this by comparing the records data_hash and deleted_at fields before and after the update
+                const existingIds = chunk.map((r) => r.external_id);
+                const updateQuery = trx
+                    .with('existing', (qb) => {
+                        qb.select('external_id', 'data_hash', 'deleted_at')
+                            .from(RECORDS_TABLE)
+                            .where({
+                                connection_id: connectionId,
+                                model
+                            })
+                            .whereIn('external_id', existingIds)
+                            .forUpdate();
+                    })
+                    .with('upsert', (qb) => {
+                        qb.insert(encryptedRecords)
+                            .into(RECORDS_TABLE)
+                            .onConflict(['connection_id', 'external_id', 'model'])
+                            .merge()
+                            .whereIn(`${RECORDS_TABLE}.external_id`, trx.select('external_id').from('existing'))
+                            .returning(['external_id', 'data_hash', 'deleted_at']);
+                    })
+                    .select<{ external_id: string; status: 'inserted' | 'changed' | 'undeleted' | 'deleted' | 'unchanged' }[]>(
+                        'upsert.external_id',
+                        trx.raw(`
+                                CASE
+                                    WHEN existing.external_id IS NULL THEN 'inserted'
+                                    ELSE
+                                        CASE
+                                            WHEN existing.deleted_at IS NOT NULL AND upsert.deleted_at IS NULL THEN 'undeleted'
+                                            WHEN existing.deleted_at IS NULL AND upsert.deleted_at IS NOT NULL THEN 'deleted'
+                                            WHEN existing.data_hash <> upsert.data_hash THEN 'changed'
+                                            ELSE 'unchanged'
+                                        END
+                                END as status`)
```

Comment:
Found this https://stackoverflow.com/a/38858662
You might already know this since it's very similar to this PR, but could simplify the update/insert query and delete could be done in a second query without the need to lock all rows

Reply:
I have tried to make `xmax` work but without success. I am always getting errors like `cannot retrieve a system column in this context`

Reply:
arg too bad, seems like it was a good solution